File: ios/LaneShadow/Views/Atoms/BottomSheetInput.swift
import LaneShadowTheme
import SwiftUI

// MARK: - BottomSheetInput Component

/**
 * BottomSheetInput component
 *
 * Text input designed for bottom sheets with proper keyboard handling
 * Following RN wrapper API from react-native/components/ui/input.tsx
 * Matrix reference: .spec/prds/native-rewrite/matrices/ui/atoms/Input.md
 *
 * ## Design Tokens Used
 * - Container height: `theme.control.minHeight` (48)
 * - Border radius: `theme.radius.xl` (16)
 * - Border width: `theme.borderWidth.thin` (1) when focused/error
 * - Background: `theme.colors.surface.default`
 * - Focus border: `theme.colors.primary.default`
 * - Error border: `theme.colors.danger.default`
 * - Text color: `theme.colors.onSurface.default`
 * - Placeholder color: `theme.colors.onSurface.subtle`
 * - Input text: `theme.type.body.md.fontSize` (16), regular weight
 * - Label: `theme.type.label.sm.fontSize` (12), medium weight, uppercase, `theme.colors.onSurface.subtle`
 * - Left icon padding: leading=`theme.space.lg` (16), trailing=`theme.space.sm` (8)
 * - Right icon padding: leading=`theme.space.sm` (8), trailing=`theme.space.lg` (16)
 * - Icon size: `theme.iconSize.small` (16) or medium (20)
 * - Input padding: horizontal=`theme.space.sm` (8), vertical=`theme.space.md` (12)
 * - Label gap: `theme.space.xs` (4)
 * - Disabled opacity: 0.5
 *
 * ## Parameters
 * - Parameters:
 *   - value: Current text value (binding)
 *   - placeholder: Optional placeholder text
 *   - label: Optional label text displayed above input
 *   - error: Whether to show error state (default: false)
 *   - editable: Whether the input is editable (default: true)
 *   - leftIcon: Optional SF Symbol name for left icon
 *   - rightIcon: Optional SF Symbol name for right icon
 *   - testID: Optional testing identifier for accessibility
 */
public struct BottomSheetInput: View {
    @Environment(\.theme) private var theme
    @FocusState private var isFocused: Bool

    @Binding private var value: String
    private let placeholder: String?
    private let label: String?
    private let error: Bool
    private let editable: Bool
    private let leftIcon: String?
    private let rightIcon: String?
    private let testID: String?

    public init(
        value: Binding<String>,
        placeholder: String? = nil,
        label: String? = nil,
        error: Bool = false,
        editable: Bool = true,
        leftIcon: String? = nil,
        rightIcon: String? = nil,
        testID: String? = nil
    ) {
        self._value = value
        self.placeholder = placeholder
        self.label = label
        self.error = error
        self.editable = editable
        self.leftIcon = leftIcon
        self.rightIcon = rightIcon
        self.testID = testID
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: theme.space.xs) {
            // Optional label above input
            if let label {
                Text(label.uppercased())
                    .font(.system(size: theme.type.label.sm.fontSize, weight: .medium))
                    .foregroundStyle(theme.colors.onSurface.subtle)
            }

            // Input container
            HStack(spacing: 0) {
                // Left icon (optional)
                if let leftIcon {
                    Image(systemName: leftIcon)
                        .font(.system(size: theme.iconSize.medium))
                        .foregroundStyle(iconColor)
                        .frame(width: theme.iconSize.medium + theme.space.sm)
                        .padding(.leading, theme.space.lg)
                        .padding(.trailing, theme.space.sm)
                }

                // Text field
                TextField(text: $value) {
                    if let placeholder {
                        Text(placeholder)
                    }
                }
                .font(.system(size: theme.type.body.md.fontSize, weight: .regular))
                .foregroundStyle(textColor)
                .focused($isFocused)
                .disabled(!editable)
                .padding(.horizontal, theme.space.sm)
                .padding(.vertical, theme.space.md)
                .frame(minHeight: theme.control.minHeight)

                // Right icon (optional)
                if let rightIcon {
                    Image(systemName: rightIcon)
                        .font(.system(size: theme.iconSize.medium))
                        .foregroundStyle(iconColor)
                        .frame(width: theme.iconSize.medium + theme.space.sm)
                        .padding(.leading, theme.space.sm)
                        .padding(.trailing, theme.space.lg)
                }

                // Right padding when no icon
                if rightIcon == nil {
                    Spacer()
                        .frame(width: theme.space.lg)
                }
            }
            .background(theme.colors.surface.default)
            .clipShape(RoundedRectangle(cornerRadius: theme.radius.xl))
            .overlay {
                // Border overlay for focus/error states
                if shouldShowBorder {
                    RoundedRectangle(cornerRadius: theme.radius.xl)
                        .stroke(borderColor, lineWidth: theme.borderWidth.thin)
                }
            }
            .opacity(editable ? 1.0 : theme.opacity.step05) // 0.5 when disabled
        }
        .accessibilityLabel(label ?? placeholder ?? "Text input")
        .accessibilityAddTraits([.isKeyboardKey])
        .if(!editable) {
            $0.accessibilityAddTraits(.notEnabled)
        }
        .accessibilityIdentifier(testID ?? "bottom-sheet-input")
    }

    // MARK: - Private Computed Properties

    private var shouldShowBorder: Bool {
        isFocused || error
    }

    private var borderColor: Color {
        if error {
            return theme.colors.danger.default
        } else if isFocused {
            return theme.colors.primary.default
        } else {
            return .clear
        }
    }

    private var textColor: Color {
        if editable {
            return theme.colors.onSurface.default
        } else {
            return theme.colors.onSurface.subtle
        }
    }

    private var iconColor: Color {
        if error {
            return theme.colors.danger.default
        } else if isFocused {
            return theme.colors.primary.default
        } else {
            return theme.colors.onSurface.subtle
        }
    }
}

// MARK: - View Conditional Extension

private extension View {
    @ViewBuilder
    func `if`<Content: View>(
        _ condition: Bool,
        transform: (Self) -> Content
    ) -> some View {
        if condition {
            transform(self)
        } else {
            self
        }
    }
}

// MARK: - Preview

#Preview("Default") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Enter text"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("With Label") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Enter text",
        label: "Label"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("With Value") {
    BottomSheetInput(
        value: .constant("Sample text"),
        placeholder: "Enter text",
        label: "Label"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("Error State") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Enter text",
        label: "Email",
        error: true
    )
    .laneShadowTheme()
    .padding()
}

#Preview("Disabled") {
    BottomSheetInput(
        value: .constant("Disabled text"),
        placeholder: "Enter text",
        label: "Label",
        editable: false
    )
    .laneShadowTheme()
    .padding()
}

#Preview("With Left Icon") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Search",
        leftIcon: "magnifyingglass"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("With Right Icon") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Password",
        rightIcon: "eye.slash"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("With Both Icons") {
    BottomSheetInput(
        value: .constant(""),
        placeholder: "Search",
        label: "Search",
        leftIcon: "magnifyingglass",
        rightIcon: "xmark.circle"
    )
    .laneShadowTheme()
    .padding()
}

#Preview("Multiple States") {
    VStack(spacing: theme.space.lg) {
        BottomSheetInput(
            value: .constant(""),
            placeholder: "Default input",
            label: "Default"
        )

        BottomSheetInput(
            value: .constant("Focused text"),
            placeholder: "Enter text",
            label: "Focused"
        )

        BottomSheetInput(
            value: .constant(""),
            placeholder: "Enter email",
            label: "Error",
            error: true
        )

        BottomSheetInput(
            value: .constant("Can't edit"),
            placeholder: "Disabled",
            label: "Disabled",
            editable: false
        )
    }
    .laneShadowTheme()
    .padding()
}

File: android/app/src/main/java/com/laneshadow/ui/components/atoms/BottomSheetInput.kt
package com.laneshadow.ui.components.atoms

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.FocusInteraction
import androidx.compose.foundation.interaction.Interaction
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.editableText
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * BottomSheetInput component props
 *
 * Following RN wrapper API from react-native/components/ui/bottom-sheet-input.tsx
 *
 * @param value Current text value in the input field
 * @param onValueChange Callback when text value changes
 * @param placeholder Placeholder text to show when input is empty
 * @param label Label text to display above the input
 * @param error Whether to show error state (red border)
 * @param editable Whether the input can be edited (default: true)
 * @param leftIcon Optional icon composable to display on the left side
 * @param rightIcon Optional icon composable to display on the right side
 * @param modifier Modifier for the container
 * @param testID Test identifier for UI testing (defaults to null)
 */
@Suppress("UNUSED_PARAMETER")
@Composable
fun BottomSheetInput(
    value: String = "",
    onValueChange: (String) -> Unit = {},
    placeholder: String? = null,
    label: String? = null,
    error: Boolean = false,
    editable: Boolean = true,
    leftIcon: @Composable (() -> Unit)? = null,
    rightIcon: @Composable (() -> Unit)? = null,
    modifier: Modifier = Modifier,
    testID: String? = null,
) {
    val theme = LocalLaneShadowTheme.current

    // Matrix constants documented against BottomSheetInput.md:
    // - Container height: size.inputHeight = 48
    // - Border radius: radius.xl = 16
    // - Border width: borderWidth.thin = 1
    // - Icon size: 20dp
    // - Input text: type.body.md.fontSize = 16
    // - Label: type.label.sm.fontSize = 12
    val INPUT_HEIGHT = 48.dp
    val ICON_SIZE = 20.dp
    val LABEL_FONT_SIZE = 12.sp
    val INPUT_FONT_SIZE = 16.sp

    var isFocused by remember { mutableStateOf(false) }
    val focusRequester = remember { FocusRequester() }
    val interactionSource = remember { MutableInteractionSource() }

    // Track focus state from interactions
    LaunchedEffect(interactionSource) {
        interactionSource.interactions.collect { interaction ->
            when (interaction) {
                is FocusInteraction.Focus -> isFocused = true
                is FocusInteraction.Unfocus -> isFocused = false
            }
        }
    }

    // Determine border color and width based on state
    val borderColor = when {
        error -> theme.colors.danger.default
        isFocused -> theme.colors.primary.default
        else -> Color.Transparent
    }
    val borderWidth = when {
        error || isFocused -> 1.dp
        else -> 0.dp
    }

    // Disabled opacity
    val containerAlpha = if (editable) 1.0f else 0.5f

    // Container with label and input
    Box(modifier = modifier) {
        if (label != null) {
            Text(
                text = label,
                style = theme.type.label.sm,
                fontSize = LABEL_FONT_SIZE,
                color = theme.colors.onSurface.default,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(bottom = theme.space.xs)
                    .semantics {
                        contentDescription = label
                    }
            )
        }

        Surface(
            modifier = Modifier
                .fillMaxWidth()
                .height(INPUT_HEIGHT)
                .alpha(containerAlpha)
                .then(
                    if (testID != null) {
                        Modifier.testTag(testID)
                    } else {
                        Modifier
                    }
                )
                .semantics {
                    contentDescription = label ?: placeholder ?: "Text input"
                    if (!editable) {
                        disabled()
                    }
                    editableText = AnnotatedString(value)
                },
            shape = RoundedCornerShape(theme.radius.xl),
            color = theme.colors.surface.default,
            border = if (borderWidth > 0.dp) {
                BorderStroke(borderWidth, borderColor)
            } else null,
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .height(INPUT_HEIGHT),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Left icon with padding
                if (leftIcon != null) {
                    Box(
                        modifier = Modifier
                            .padding(start = theme.space.lg, end = theme.space.sm)
                            .width(ICON_SIZE)
                            .height(ICON_SIZE),
                        contentAlignment = Alignment.Center
                    ) {
                        leftIcon()
                    }
                } else if (leftIcon == null && rightIcon != null) {
                    // Add spacing to align with inputs that have icons
                    Spacer(modifier = Modifier.width(theme.space.lg))
                }

                // Text input field
                BasicTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier
                        .weight(1f)
                        .focusRequester(focusRequester)
                        .padding(horizontal = theme.space.sm, vertical = theme.space.md),
                    enabled = editable,
                    singleLine = true,
                    textStyle = TextStyle(
                        color = theme.colors.onSurface.default,
                        fontSize = INPUT_FONT_SIZE,
                    ),
                    cursorBrush = SolidColor(theme.colors.primary.default),
                    keyboardOptions = KeyboardOptions.Default.copy(
                        keyboardType = KeyboardType.Text
                    ),
                    keyboardActions = KeyboardActions.Default,
                    interactionSource = interactionSource,
                )

                // Right icon with padding
                if (rightIcon != null) {
                    Box(
                        modifier = Modifier
                            .padding(start = theme.space.sm, end = theme.space.lg)
                            .width(ICON_SIZE)
                            .height(ICON_SIZE),
                        contentAlignment = Alignment.Center
                    ) {
                        rightIcon()
                    }
                } else if (rightIcon == null && leftIcon != null) {
                    // Add spacing to align with inputs that have icons
                    Spacer(modifier = Modifier.width(theme.space.lg))
                }
            }
        }
    }
}
